### DISCUSSION TITLE
dynamic 1d pipe model with heat exchange

### DISCUSSION THREAD

**Discussion by atpeterz** [IDAES/idaes-pse#1116] [Q&A]
![image](https://user-images.githubusercontent.com/127108534/223109020-fe6cada2-a951-45fe-b3c7-089c682d063b.png)

**Comment by adowling2** [comment]
@atpeterz Welcome to the IDAES discussion board. I will let another member of the IDAES team speak to your specific questions about a 1DControlVolume.

As a fallback, if you are only interested in quickly simulating the PDEs you posted, you can try using Pyomo directly. Here are two relevant examples:
https://jckantor.github.io/ND-Pyomo-Cookbook/notebooks/05.03-Heat_Conduction_in_Various_Geometries.html
https://jckantor.github.io/ND-Pyomo-Cookbook/notebooks/05.04-Diffusion_Adsorption_in_Polymers.html

If the dynamic 1D pipe model is just a starting point for a larger project, I encourage you to try to simulate this system with IDAES. IDAES will allow you to build more complex systems by coupling various unit operations from the modeling library.

**Comment by andrewlee94** [comment]
@atpeterz Welcome to the IDAES discussion board. It is definitely possible to solve the types of models you are interested in IDAES, although dynamic models are always a little bit tricky to work with. Unfortunately, we haven;t had the time or resources to put together good examples of working with dynamic models yet, but we have solved dynamic systems, including heat exchangers, before.

As a place to start, I would recommend looking at the 1D heat exchanger model - the case you discussed is similar in many ways to this: https://github.com/IDAES/idaes-pse/blob/main/idaes/models/unit_models/heat_exchanger_1D.py.

Regarding control volume,s the first thing to emphasize is that these are meant to be be convenience tools which can automate common tasks/activities for you - you do not need to use them if you do not want to. For example, the second equation you showed appears to be the energy balance for the fluid phase with a heat transfer term - the control volume can automate the construction of that for you. The `add_total_energy_balance` method can add the constraint `V*d(rho*h)/dt = - dF/dx + Q` for you - you just need to provide a constraint that defines `Q=U*A*deltaT` somewhere else.

Next, the place to implement your additional equations (the wall material energy balance and the definition of Q) is either the unit model or the control volume (we would generally use the unit model). Note that an instance StateBlock is created at every point in the spatial domain, so you do not want to have any derivative variables in there (you would end up with n-instance of the derivative, rather than a single instance with n-elements).

Finally, a comment on general solvability of these types of systems: the system you showed above does not seem to be too bad, but you do have two different time-dependent derivatives which quite likely have different time constants (fluid and wall hold-ups). If these time constants are significantly different then the system of equations will become very stiff and difficult to solve. If you have a case where the time constants are extremely different, you generally need to decide which of the two terms is more important and simplify the other (generally by assuming steady-state over the time interval of interest). We have tried solving these types of systems before, and we found htat we needed to make these sorts of assumptions in order to be able to solve the model.

**Reply by atpeterz** [reply]
@andrewlee94 Thank you for your fast reply. That made things a bit more clear for me. 

I searched for the `add_total_energy_balance`, but does it exist anymore? In the documentation it says that the method is no longer supported. Could I also use instead the add_total_enthalpy_balance method?

As already stated in my question above, i started with a single 1DControlVolume. I also wrote my own property package with an own initialization routine. I don´t know if I really need this for such a single application but it was stated in the cutom prpoerty packages documentaiton of IDAES. The code is mainly taken from the custom_physical_property_package tutorial in the example repo. Also I don´t know if it works correct, since all variables are fixed ( I think after initialization). 
Generally where is the difference from the initialization and realease_state methods inside the StateBlock Methods class and the one in the ControlVolume/Unit Model?

At least my programme runs with a dynamic Flowsheet `m.fs = FlowsheetBlock(dynamic = True, time_units = ...)`, but only in theory since all inlet proerties are fixed.

Could you please have a look at the code and show me how I could make the inlet temperature dynamic (e.g. a step in temperature at a specific timepoint) to be able to get information about flow time of the fluid through the control volume? 
[1d_customunit.zip](https://github.com/IDAES/idaes-pse/files/10909662/1d_customunit.zip)

Thanks in advance!

**Reply by andrewlee94** [reply]
Sorry - I meant to say `total_enthalpy_balance` (the total energy balance option was added as a future option we haven't ended up needing yet).

IDAES is designed in a hierarchical structure, and a lot of the decision making is pushed down to the property package level. Generally speaking, you need a property package suitable for your application along with its own initialization routine (however those initialization routines are often very simple - in some cases just fixing the state variables, solving and then unfixing the state variables).

The initialization method in the control volume is very simple, and by itself is does not do much; it is intended to be a helper to do common steps at the unit level. Basically, the 1D control volume does the following:

1. Fix the state variables at all time and space points if they are not fixed already. The values are either fixed to the initial guesses you provide in the `state_args` argument, or to the current state at t=0, x=0 if no initial guesses are provided.
2. Call `the initialization method of the state block (cv.properties.initialize()).
3. If `hold_state` was set to `False`, unfix all the states fixed in step 1, other wise return a dict containing what was fixed so that the unit model can later call `release_state()` passing the dict of what needs to be unfixed.

Note that if everything is done correctly, all the states should be in the same state they started in once initialization is completed (i.e, fixed or unfixed).

I'll try to find some time to take a look at your code later this week.

**Reply by atpeterz** [reply]
Thank you for your explanations. I am looking forward for your reply regarding the code. I updated it a little bit, here is the new version. It works but the output values are not what i expected them to be. The output temperature should be the same as at the inlet, actually it is ~600 K higher in my model. But their are no operations going on, just the Control Volume. Also the Initialization is really slow and fails if a step of temperature is given at the input (e.g. after 20 seconds ). I don´t know if their is something wrong with the configuration of the model or if the initialization isn´t working correct. Thank you very much for your help!

[1d_customunit.zip](https://github.com/IDAES/idaes-pse/files/10921253/1d_customunit.zip)

Thank you very much for your help!

**Reply by andrewlee94** [reply]
I had a fairly quick look through the code, and I can see a few big things that need to be addressed:

1. You need to call a DAE transformation from pyomo.dae, otherwise the time derivatives are not expanded to link each time point. Also note that this will automatically add time points to the time domain up to the number you specify in the transformation, so you do not need to provide a full set of point when you crate the domain. See https://pyomo.readthedocs.io/en/latest/modeling_extensions/dae.html - you want to use a forward finite difference method (or maybe a collocation method).
2. I do not see you adding a momentum/pressure balance to the control volume. This will mean that pressure will not be constrained by a balance, leaving additional degrees of freedom that you would need to deal with.
3. Building from this - make sure you have exactly 0 degrees of freedom when you initialize the model (I didn't try running it, but I suspect you either have degrees of freedom or you have had to specify things that you would not normally need to). Initializing with degrees of freedom means that you have no idea what the final solution might be as things can move around freely.
4. Seeing as the properties have only water, I would suggest dropping mole_frac_comp from the model - it will always be one so it is not actually needed anywhere (at least for this case). Having unnecessary variables makes solving the model harder.

**Reply by atpeterz** [reply]
Hi @andrewlee94, thanks again.

I tried to implement your instructions and tips into my code but yet their are still some issues I am struggling with:

What do you mean with the first point? I thought it would be enough for calling the DAE transformation in the ControlVolume1DBlock? ..like here 
![image](https://user-images.githubusercontent.com/127108534/224543447-05f17f30-45b5-454b-82bb-6ee43c48c47c.png)

and `control_volume.apply_transformation()`

 with the configurations given in the config block in my unit model.

![image](https://user-images.githubusercontent.com/127108534/224543261-4f6b6242-9314-48e4-ace5-f376b640f8a9.png)

Do I need to set any other configurations for the dynamic Model?

**Reply by atpeterz** [reply]
I managed to get zero degrees of freedom for the initialization of a steady state system, where I fixed pressure and temperature for all times and spaces, but flow_mol only for the inlet and outlet; And no fixing of any geometric constraints.

But when I run the same model in dynamic mode, I don´t get zero degrees of freedom. Also not if I fix also flow_mol for all times and all spaces. Are their any other variables I need to fix in order to initialize the model correct?  
I also tried solving with fixing material hold up for initialization, but that also gave wrong results. I am struggling now since two days at this point and don´t know what s going wrong here. I also added momentum balance to the unit model.
I haven´t declared any other state variables as pressure, temperature and flow_mol. Is their anything else to be considered working with dynamic models in IDAES? Here is my try of fixing the states for the initialization. 

![image](https://user-images.githubusercontent.com/127108534/224547383-10457297-bd65-4446-8a5d-3916d8eded8b.png)

Here are also the updated files with the code if you need it.
[1d_flow_model.zip](https://github.com/IDAES/idaes-pse/files/10951082/1d_flow_model.zip)

Thank you and kind regards,
Peter

**Reply by andrewlee94** [reply]
Hi again,

For the first point, you have two separate continuous domains in the system - one for space and one for time - and each of those need to be transformed separately. The control volume does the spatial transformation for you, but you still need to do the time domain yourself. This is because the spatial domain is limited to a single unit but the time domain covers the full flowsheet (and thus needs to be left until after all units have been defined).

For the second point, your issues are partial due to not having transformed the time domain - without doing that you essentially have multiple separate steady state models with extra degrees of freedom. You definitely need the momentum balance however, as pressure is critical in dynamic models to determine the holdup at each point. One you have transformed the time domain you will find you do not need to fix most of the things you are fixing now - you should just need the inlet conditions and some pressure-flow relationship at all points in the model. The pressure-flow relationship is required to couple holdup (i.e. accumulation) to the material flow as otherwise you will have an extra degree of freedom.

**Reply by atpeterz** [reply]
Ok, we are getting closer to my desired result..

I implemented now the transformation for the domain, and it lowered the dof as you told me.
But yet I don´t get what i want:
I added the momentum_balance to the control volume;
![image](https://user-images.githubusercontent.com/127108534/224814730-2cdb6736-d1ee-4f77-bc3d-7cb8b804bcd8.png)
in the config Block the balance type is declared like here 
![image](https://user-images.githubusercontent.com/127108534/224821615-929b89a7-c9d5-4925-bfd0-d6264fa5b4de.png)
Do I need to consider anything else for configuring the momentum balance?

For Initialization I fixed now the inlet of temperature, pressure and flow_mol. 
Regarding the pressure-flow relationship I also fixed deltaP ( `m.fs.odf.control_volume.deltaP.fix()` ) and flow_mol for all time and space points, since these should have a direct relation (regarding info from the internet). But I am not sure if that is true.
I also fixed length and area for the initialization. 
That led my model to dof = 4 for initialization, with the configuration given here (9 timepoints and 3 space points)
![image](https://user-images.githubusercontent.com/127108534/224819948-baf6917c-4cf5-46ae-8cc8-126306e59e68.png)


After initialization i unfixed everything except the inlet conditions for temperature, pressure and flow_mol and the geometry constraints for length and area.
This is my result at the input:
![image](https://user-images.githubusercontent.com/127108534/224820421-e4933065-599a-4fc9-bdbd-9a0873edb97c.png)

The Flow_mol stays now constant except for the last time point and the temperature is always equals the temperature_ref value that is given in the property package. It should be always equals 300 K (constantly given at the inlet for all times)  When I choose the time discretization scheme=BACKWARD, the first timepoint gives always the wrong flow_mol value. 

I also tried fixing material holdup, but that also led to a false result.

Sry for having so many questions for such a small problem, this is my first time working with pyomo and IDAES. Anyway, thanks for your answers.

Kind regards
Peter

**Reply by andrewlee94** [reply]
@atpeterz You are getting close, but I think you might need to take a step back and think about the equations you need to write for this system first, and then start mapping these to equations/methods in IDAES and Pyomo. It is also generally a good idea to start with a steady-state model first, and then move to dynamics, as the steady-state model always forms the basis of the dynamic model. I.e., when you move to dynamics all the things you did in the steady-state model still apply, but you need to add more information.

Firstly, I doubt you want to fix the molar flow in the pipe - if you ado that then there is no need for a model any more as you know the flow rates. When people say there is a direct relationship between deltaP and flow, they mean an equation of the form `deltaP = k*flow`; i.e., that flow rate is proportional to pressure drop. This is the simplest form of pressure-flow relationship.

The other thing I think you are missing are the initial conditions for the system. When you move to a dynamic model, one of the things you need to provide is information on the accumulation/holdup the system at time 0. One common starting point is assuming steady-state, in which case you fix all the time derivatives to zero at t=0 (but only t=0). This is needed because if you think about the material balance you have an extra degree of freedom in the accumulation term that needs to be described.

**Reply by atpeterz** [reply]
Well.. I thought the equations for energy, material and momentum balance are automatically implemented in the several add_..._balance methods that are provided by the Control Volume? Does that mean I have to choose and add a constraint for the pressure-Flow relationship on my own? I would have assumed that the add_momentum_balance method does that automatically for me.

**Reply by andrewlee94** [reply]
Those methods add the conservation equations, but dynamic models need an additional equation. The reason is not the momentum/pressure balance either, as that is fairly well defined, but instead due to the accumulation/holdup term in the material balance.

Consider that in a steady state model you have `0 = Flow In-Flow Out+Generation`, `0 = Energy In - Energy out + Heat + Work` and `0 = Pressure In - Pressure Out + deltaP` and the model is well defined. When you go to a dynamic system, you add a new accumulation term to the material and energy balances, thus an extra degree of freedom which needs an additional equation.

Further, the pressure-flow relationship is very case specific; thee is no single form that is commonly applicable, thus we need to defer that to the user to decide.

**Reply by atpeterz** [reply]
Thank you!


I added the additional equation for the dynamic model (the pressure-Flow relation) in the build method inside the unit model, its of the form Flow= const. deltaP, as this is the easiest.
![image](https://user-images.githubusercontent.com/127108534/225968443-937b290d-a782-42a6-b9fc-15f5830eeed0.png)

For the initialization I fixed temperatures at all time and space points, Flow_mol for inlet and outlet for all times and pressure only for inlet and outlet at time=0
![image](https://user-images.githubusercontent.com/127108534/225976292-0ec7928a-b4e5-4e19-992b-7590c0059079.png)

You told me I should also fix all time derivatives, so additionally I fixed also enthalpy_flow_dx and material_flow_dx for t=0 and set it to 0.
There is pressure_dx as a derivative variable also, but then my final dof would be <0 so I didn´t do anything with that.
![image](https://user-images.githubusercontent.com/127108534/225976406-765cc35f-b3b9-4320-bee1-d62397a9d930.png)

That led my model to dof = 0 for initialization.

After initialization i unfixed all temperatures except for t=0, und also all outlet conditions of flow and pressure.
And for final solve I also fixed area and length of the control volume.
![image](https://user-images.githubusercontent.com/127108534/225976735-607ccbc3-e04f-496f-abf3-f448f4cf36ae.png)


So for initialization I defined all temperatures. 
I am explaining this so detailed, because I would have imagined that I would get a realistic result. But when the model is simulated for longer times, one sees that the temperatures at the output relax back to the value defined in the property package for the reference temperature (=298.15 K) that is used for calculation of enthalpy and the `get_energy_density_terms` method. 

Do you have a explanation how this is possible, because before initialization I defined and fixed temperature values to all available states and as I have a constant inlet temp. of 310 K, their should not be much change of outlet temperature, because for initialization everything was set = 310 K. I am also wondering why for the last timepoint the flow rate always is false, this behaviour was the same for all my test runs yet.

Here is the result I got from the solver:
[reult.txt](https://github.com/IDAES/idaes-pse/files/11004117/reult.txt)

I am also wondering about the high number of dof the solver has to deal with, but since I have to unfix temperatures I didn´t know what else to do. Is that the standard?

Its really quite tricky to develop dynamic models in IDAES, even when their is just a simple 1d flow...

**Reply by andrewlee94** [reply]
@atpeterz Modelling dynamic systems is never easy - even 1D flow is harder than you might think.

I strongly suggest you start by modeling the steady-state system and get that right first; a steady-state model is always the foundation of the dynamic model. This will also help guide you to what degrees of freedom you should be fixing. Once the steady-state model is working, then switch to dynamics and look at how many additional degrees of freedom you get - these degrees of freedom are all related to dynamic phenomena, i.e. time derivative or initial conditions.

The thing that immediately jumped out at me in you description was this:

> For the initialization I fixed temperatures at all time and space points, Flow_mol for inlet and outlet for all times and pressure only for inlet and outlet at time=0

Those are almost definitely not the things you want to be fixing, and doing a steady-state model would probably show you that.

1. If you are fixing the flow rate at both the inlet and outlet, then what dynamics are actually occurring? You are implicitly assuming in-compressible flow with no disturbances, but if you are using the IAPWS property package (which I think you were), then this is not correct.
2. Similarly, fixing pressure at the inlet and outlet just constrained you pressure profile - what are you actually solving for now?
3. Finally, why are you fixing temperature at all points in space and time? Combined with the above, there are now no dynamic at all in the system.

Even if this is for just initialization, you need to set your initial conditions to actually reflect a real system. Otherwise, you end up with a degenerate model where parts of the system are over constrained and others are under constrained. I suspect that is why you are seeing unexpected temperature values - there is nothing pushing them to take meaningful value anymore.

**Reply by atpeterz** [reply]
@andrewlee94 Thank you for your quick reply

Well I thought that my steady-state model already works correct - at least it had the output that i expected: a temp inside the controlvolume rising from 300 up to 350 K, or anything that is set as start and end temperature; This was the reason I thought I could move on to dynamic modelling.

But I fixed for steady state all pressures and temperatures for initialization. I thought the pressure fixing was needed for the add_momentum_balance_type.totalpressure method and the temperatures to tell the model where i wanted them to have (temperature rise from inlet up to outlet). And I fixed flow at inlet and outlet for the material balance, but only for initialization. For the final solve everything expect the inlet for all three state variables and the outlet of temperature was unfixed again. 
Since the modeled temperatures had very small deviations from the initialized ones I assumed the model solved correct.

I am using my own written property package, not the IAPWS properties.

I thought that the model is actually solved after the initialization. Therefore I assumed its no problem to fix temperatures everywhere since they are anyway unfixed for the final solve. But your last argument  makes also sense, thanks for the explanations!

**Reply by atpeterz** [reply]
regarding your point 1) from last reply: Here I am considering an in-compressible fluid.  For the beginning, I wanted nothing more than when having a constant flow into the pipe with changing temperatures to model the change of temperatures inside the pipe. So to speak, at the outlet I wanted to see the same "temperature curve" as given at the inlet, only shifted in time. This was my thinking of dynamic behaviour of the starting model, for later being able to model the heatexchange with environment (the 2 coupled PDEs given in my opening question of this thread).

**Reply by andrewlee94** [reply]
The problem is with the way you implemented it however - if the system is in-compressible, then the mass balances should ensure that flow in equals flow out at all times. Thus, you should not be fixing the outlet flow rate. What you have ended up doing is creating a degenerate system, where the degrees of freedom appear to be zero overall but in reality part of the system is over constraint and part is under constrained. In short, this means that you have a system with an infinite number of solutions; importantly this means that a solver can still find a solution to the problem but the solution it is essentially meaningless.

In short - I suspect that your steady-state model is incorrect even before you consider the dynamic aspect.

**Reply by atpeterz** [reply]
yeah i am back working an the steady state model. With the pressure Flow relation included also their I don´t need to fix outlet conditions except for temperature any more. But then i am still left with 3 dof, when I have 5 space points at all (finite elements = 4).
To me it makes sense that I have to fix the temperatures also inside the cv, because I haven´t defined a constraint for temperatures. It is only occuring in the get_energy_density terms and get_enthalpy_flow_terms methods, that are defined in the stateblockdata class in the prop package.

I will try my best to make it working

**Reply by andrewlee94** [reply]
You should not need to fix outlet temperature either - if you have to fix that then it means you are not solving for temperature in the system. Temperature should be coupled to enthalpy and thus be defined by the energy balance equations.

However, looking at the code you posted earlier, it looks like you are including heat transfer terms in your model - I suspect these are your extra degrees of freedom. You need to provide some constraint on heat transfer at each finite element in order to define the model, otherwise how does the solver know how much heat should be added or removed at each stage? This would also explain why you saw weird temperature behaviour, as there is nothing to force the temperature to move away from its default value (rather than change the temperature and disturb everything else, the solver can just set the heat transfer to whatever is necessary to keep temperature whee it is).
